rootfs/etc/nginx/lua/balancer/leastconn.lua

```diff
@@ -0,0 +1,108 @@
+local util = require("util")
+local split = require("util.split")
+require("resty.core")
+
+local ngx = ngx
+local ipairs = ipairs
+local setmetatable = setmetatable
+local string_format = string.format
+local ngx_log = ngx.log
+local DEBUG = ngx.DEBUG
+
+
+local _M = { name = "leastconn" }
+
+function _M.new(self, backend)
+    local o = {
+        peers = backend.endpoints
+    }
+    setmetatable(o, self)
+    self.__index = self
+    return o
+end
+
+function _M.is_affinitized()
+    return false
+end
+
+
+local function get_upstream_name(upstream)
+    return upstream.address .. ":" .. upstream.port
+end
+
+
+function _M.balance(self)
+    local peers = self.peers
+    local endpoint = peers[1]
+    local endpoints = ngx.shared.balancer_leastconn
+    local feasible_endpoints = {}
+
+    if #peers ~= 1 then
+        local lowestconns = 2147483647
+        -- find the lowest connection count
+        for _, peer in pairs(peers) do
+            local peer_name = get_upstream_name(peer)
+            local peer_conn_count = endpoints:get(peer_name)
+            if peer_conn_count  == nil then
+                -- Peer has never been recorded as having connections - add it to the connection
+                -- tracking table and the list of feasible peers
+                endpoints:set(peer_name,0,0)
```

Comment:
I'm surprised we don't have a default expiry on these entries. Does it make sense to add a long expiry (2 days) by default, just to catch any leaks?